## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#074154d5-4cb7-4622-a64e-72e5001631ab) |
| mint-gateway-azure.sh | :x: [more...](#361c9863-3a12-40dd-8b51-d9f2169c2289) |
<a name="361c9863-3a12-40dd-8b51-d9f2169c2289"></a>
### 5428-7100e40/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32143
ACCESS_KEY:      minioazure1
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp fb0b36c5e63a:/mint/log /tmp/mint-logs'
(1/11) Running aws-sdk-go tests ... done in 3 seconds
(2/11) Running aws-sdk-php tests ... done in 2 minutes and 12 seconds
(3/11) Running aws-sdk-ruby tests ... done in 49 seconds
(4/11) Running awscli tests ... done in 7 minutes and 26 seconds
(5/11) Running mc tests ... done in 7 minutes and 1 seconds
(6/11) Running minio-dotnet tests ... done in 1 minutes and 5 seconds
(7/11) Running minio-go tests ... done in 2 seconds
(8/11) Running minio-java tests ... done in 45 seconds
(9/11) Running minio-js tests ... done in 1 minutes and 53 seconds
(10/11) Running minio-py tests ... FAILED in 1 minutes and 2 seconds
{
  "status": "FAIL",
  "args": {
    "object_name": "d43ca9ec-0b99-4450-8085-c4a521fcda9b",
    "bucket_name": "minio-py-test-70f9dad8-8948-4996-a39e-892a2d19ba34"
  },
  "duration": 1232,
  "message": "Listed upload ids do not match the created ids",
  "name": "minio-py:test_list_incomplete_uploads",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1127, in test_list_incomplete_uploads\n    raise ValueError(\"Listed upload ids do not match the created ids\")\nValueError: Listed upload ids do not match the created ids\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1807, in main\n    test_list_incomplete_uploads(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1129, in test_list_incomplete_uploads\n    raise Exception(err)\nException: Listed upload ids do not match the created ids\n",
  "function": "list_incomplete_uploads(bucket_name, prefix, recursive)"
}

Executed 9 out of 11 tests successfully.
```
<a name="074154d5-4cb7-4622-a64e-72e5001631ab"></a>
### 5428-7100e40/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30856
ACCESS_KEY:      AKIAJH7GQKYNU3C2ADXA
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ce649618a22d:/mint/log /tmp/mint-logs'
(1/11) Running aws-sdk-go tests ... done in 2 seconds
(2/11) Running aws-sdk-php tests ... done in 1 minutes and 18 seconds
(3/11) Running aws-sdk-ruby tests ... done in 1 minutes and 50 seconds
(4/11) Running awscli tests ... done in 8 minutes and 6 seconds
(5/11) Running mc tests ... done in 6 minutes and 18 seconds
(6/11) Running minio-dotnet tests ... done in 1 minutes and 43 seconds
(7/11) Running minio-go tests ... done in 18 seconds
(8/11) Running minio-java tests ... FAILED in 51 seconds
{
  "name": "minio-java",
  "function": "setBucketPolicy(String bucketName, String objectPrefix, PolicyType policyType)",
  "duration": 6548,
  "status": "FAIL",
  "error": "error occured\nErrorResponse(code=AccessDenied, message=Access Denied., bucketName=null, objectName=null, resource=/minio-java-test-2u0p4m/set-bucket-policy-read-write/minio-java-test-2hqpb26, requestId=3L137, hostId=3L137)\nrequest={method=GET, url=http://minikube:30856/minio-java-test-2u0p4m/set-bucket-policy-read-write/minio-java-test-2hqpb26, headers=Host: minikube:30856\nUser-Agent: Minio (amd64; amd64) minio-java/dev\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20180118T205126Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20180118/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=403, headers=Accept-Ranges: bytes\nContent-Type: application/xml\nServer: Minio/DEVELOPMENT.2018-01-18T20-27-31Z (linux; amd64)\nVary: Origin\nX-Amz-Request-Id: 150B02B77BB0C638\nDate: Thu, 18 Jan 2018 20:51:26 GMT\nTransfer-Encoding: chunked\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1205), io.minio.MinioClient.executeGet(MinioClient.java:1298), io.minio.MinioClient.getObject(MinioClient.java:1606), io.minio.MinioClient.getObject(MinioClient.java:1510), FunctionalTest.testSetBucketPolicy(FunctionalTest.java:2427), FunctionalTest.setBucketPolicy_test4(FunctionalTest.java:2482), FunctionalTest.runQuickTests(FunctionalTest.java:2788), FunctionalTest.main(FunctionalTest.java:2830)]"
}

Executed 7 out of 11 tests successfully.
```

--- a/cmd/lock-rpc-server.go
+++ b/cmd/lock-rpc-server.go
@@ -19,7 +19,6 @@ package cmd
 import (
 	"fmt"
 	"math/rand"
-	"path"
 	"sync"
 	"time"
 
@@ -64,49 +63,45 @@ type lockServer struct {
 }
 
 // Start lock maintenance from all lock servers.
-func startLockMaintenance(lockServers []*lockServer) {
-	for _, locker := range lockServers {
-		// Start loop for stale lock maintenance
-		go func(lk *lockServer) {
-			// Initialize a new ticker with a minute between each ticks.
-			ticker := time.NewTicker(lockMaintenanceInterval)
+func startLockMaintenance(lkSrv *lockServer) {
+	// Start loop for stale lock maintenance
+	go func(lk *lockServer) {
+		// Initialize a new ticker with a minute between each ticks.
+		ticker := time.NewTicker(lockMaintenanceInterval)
 
-			// Start with random sleep time, so as to avoid "synchronous checks" between servers
-			time.Sleep(time.Duration(rand.Float64() * float64(lockMaintenanceInterval)))
-			for {
-				// Verifies every minute for locks held more than 2minutes.
-				select {
-				case <-globalServiceDoneCh:
-					// Stop the timer upon service closure and cleanup the go-routine.
-					ticker.Stop()
-					return
-				case <-ticker.C:
-					lk.lockMaintenance(lockValidityCheckInterval)
-				}
+		// Start with random sleep time, so as to avoid "synchronous checks" between servers
+		time.Sleep(time.Duration(rand.Float64() * float64(lockMaintenanceInterval)))
+		for {
+			// Verifies every minute for locks held more than 2minutes.
+			select {
+			case <-globalServiceDoneCh:
+				// Stop the timer upon service closure and cleanup the go-routine.
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				lk.lockMaintenance(lockValidityCheckInterval)
 			}
-		}(locker)
-	}
+		}
+	}(lkSrv)
 }
 
 // Register distributed NS lock handlers.
 func registerDistNSLockRouter(mux *router.Router, endpoints EndpointList) error {
 	// Start lock maintenance from all lock servers.
-	startLockMaintenance(globalLockServers)
+	startLockMaintenance(globalLockServer)
 
 	// Register initialized lock servers to their respective rpc endpoints.
-	return registerStorageLockers(mux, globalLockServers)
+	return registerStorageLockers(mux, globalLockServer)
 }
 
 // registerStorageLockers - register locker rpc handlers for net/rpc library clients
-func registerStorageLockers(mux *router.Router, lockServers []*lockServer) error {
-	for _, lockServer := range lockServers {
-		lockRPCServer := newRPCServer()
-		if err := lockRPCServer.RegisterName(lockServiceName, lockServer); err != nil {
-			return errors.Trace(err)
-		}
-		lockRouter := mux.PathPrefix(minioReservedBucketPath).Subrouter()
-		lockRouter.Path(path.Join(lockServicePath, lockServer.ll.serviceEndpoint)).Handler(lockRPCServer)
+func registerStorageLockers(mux *router.Router, lkSrv *lockServer) error {
+	lockRPCServer := newRPCServer()
+	if err := lockRPCServer.RegisterName(lockServiceName, lkSrv); err != nil {
+		return errors.Trace(err)
 	}
+	lockRouter := mux.PathPrefix(minioReservedBucketPath).Subrouter()
+	lockRouter.Path(lockServicePath).Handler(lockRPCServer)
 	return nil
 }
 

--- a/cmd/lock-rpc-server_test.go
+++ b/cmd/lock-rpc-server_test.go
@@ -425,8 +425,8 @@ func TestLockRpcServerExpired(t *testing.T) {
 	}
 }
 
-// Test initialization of lock servers.
-func TestLockServers(t *testing.T) {
+// Test initialization of lock server.
+func TestLockServerInit(t *testing.T) {
 	if runtime.GOOS == globalWindowsOSName {
 		return
 	}
@@ -438,8 +438,10 @@ func TestLockServers(t *testing.T) {
 	defer os.RemoveAll(rootPath)
 
 	currentIsDistXL := globalIsDistXL
+	currentLockServer := globalLockServer
 	defer func() {
 		globalIsDistXL = currentIsDistXL
+		globalLockServer = currentLockServer
 	}()
 
 	case1Endpoints := mustGetNewEndpointList(
@@ -468,26 +470,27 @@ func TestLockServers(t *testing.T) {
 
 	globalMinioHost = ""
 	testCases := []struct {
-		isDistXL         bool
-		endpoints        EndpointList
-		totalLockServers int
+		isDistXL  bool
+		endpoints EndpointList
 	}{
 		// Test - 1 one lock server initialized.
-		{true, case1Endpoints, 1},
-		// Test - 2 two servers possible.
-		{true, case2Endpoints, 2},
+		{true, case1Endpoints},
+		// Test - similar endpoint hosts should
+		// converge to single lock server
+		// initialized.
+		{true, case2Endpoints},
 	}
 
 	// Validates lock server initialization.
 	for i, testCase := range testCases {
 		globalIsDistXL = testCase.isDistXL
-		globalLockServers = nil
+		globalLockServer = nil
 		_, _ = newDsyncNodes(testCase.endpoints)
 		if err != nil {
 			t.Fatalf("Got unexpected error initializing lock servers: %v", err)
 		}
-		if len(globalLockServers) != testCase.totalLockServers {
-			t.Fatalf("Test %d: Expected total %d, got %d", i+1, testCase.totalLockServers, len(globalLockServers))
+		if globalLockServer == nil && testCase.isDistXL {
+			t.Errorf("Test %d: Expected initialized lockServer, but got uninitialized", i+1)
 		}
 	}
 }

--- a/cmd/namespace-lock.go
+++ b/cmd/namespace-lock.go
@@ -28,13 +28,17 @@ import (
 
 	"github.com/minio/dsync"
 	"github.com/minio/lsync"
+	"github.com/minio/minio-go/pkg/set"
 )
 
 // Global name space lock.
 var globalNSMutex *nsLockMap
 
-// Global lock servers
-var globalLockServers []*lockServer
+// Global lock server one per server.
+var globalLockServer *lockServer
+
+// Instance of dsync for distributed clients.
+var globalDsync *dsync.Dsync
 
 // RWLocker - locker interface to introduce GetRLock, RUnlock.
 type RWLocker interface {
@@ -56,39 +60,41 @@ type RWLockerSync interface {
 // Returns lock clients and the node index for the current server.
 func newDsyncNodes(endpoints EndpointList) (clnts []dsync.NetLocker, myNode int) {
 	cred := globalServerConfig.GetCredential()
-	clnts = make([]dsync.NetLocker, len(endpoints))
 	myNode = -1
-	for index, endpoint := range endpoints {
+	seenHosts := set.NewStringSet()
+	for _, endpoint := range endpoints {
+		if seenHosts.Contains(endpoint.Host) {
+			continue
+		}
+		seenHosts.Add(endpoint.Host)
 		if !endpoint.IsLocal {
 			// For a remote endpoints setup a lock RPC client.
-			clnts[index] = newLockRPCClient(authConfig{
+			clnts = append(clnts, newLockRPCClient(authConfig{
 				accessKey:       cred.AccessKey,
 				secretKey:       cred.SecretKey,
 				serverAddr:      endpoint.Host,
 				secureConn:      globalIsSSL,
-				serviceEndpoint: pathutil.Join(minioReservedBucketPath, lockServicePath, endpoint.Path),
+				serviceEndpoint: pathutil.Join(minioReservedBucketPath, lockServicePath),
 				serviceName:     lockServiceName,
-			})
+			}))
 			continue
 		}
 
 		// Local endpoint
-		if myNode == -1 {
-			myNode = index
-		}
+		myNode = len(clnts)
+
 		// For a local endpoint, setup a local lock server to
 		// avoid network requests.
 		localLockServer := lockServer{
 			AuthRPCServer: AuthRPCServer{},
 			ll: localLocker{
-				mutex:           sync.Mutex{},
-				serviceEndpoint: endpoint.Path,
 				serverAddr:      endpoint.Host,
+				serviceEndpoint: pathutil.Join(minioReservedBucketPath, lockServicePath),
 				lockMap:         make(map[string][]lockRequesterInfo),
 			},
 		}
-		globalLockServers = append(globalLockServers, &localLockServer)
-		clnts[index] = &(localLockServer.ll)
+		globalLockServer = &localLockServer
+		clnts = append(clnts, &(localLockServer.ll))
 	}
 
 	return clnts, myNode
@@ -149,7 +155,7 @@ func (n *nsLockMap) lock(volume, path string, lockSource, opsID string, readLock
 		nsLk = &nsLock{
 			RWLockerSync: func() RWLockerSync {
 				if n.isDistXL {
-					return dsync.NewDRWMutex(pathJoin(volume, path))
+					return dsync.NewDRWMutex(pathJoin(volume, path), globalDsync)
 				}
 				return &lsync.LRWMutex{}
 			}(),
@@ -303,7 +309,7 @@ func (n *nsLockMap) ForceUnlock(volume, path string) {
 	//   are blocking can now proceed as normal and any new locks will also
 	//   participate normally.
 	if n.isDistXL { // For distributed mode, broadcast ForceUnlock message.
-		dsync.NewDRWMutex(pathJoin(volume, path)).ForceUnlock()
+		dsync.NewDRWMutex(pathJoin(volume, path), globalDsync).ForceUnlock()
 	}
 
 	param := nsParam{volume, path}

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -186,8 +186,8 @@ func serverMain(ctx *cli.Context) {
 
 	// Set nodes for dsync for distributed setup.
 	if globalIsDistXL {
-		clnts, myNode := newDsyncNodes(globalEndpoints)
-		fatalIf(dsync.Init(clnts, myNode), "Unable to initialize distributed locking clients")
+		globalDsync, err = dsync.New(newDsyncNodes(globalEndpoints))
+		fatalIf(err, "Unable to initialize distributed locking clients")
 	}
 
 	// Initialize name space lock.

